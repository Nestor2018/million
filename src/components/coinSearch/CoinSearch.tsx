import React, { useState, useCallback, memo } from 'react';
import { View, StyleSheet, TextInput, Platform } from 'react-native';

import colors from '../../utils/colors';

interface CoinSearchProps {
  onChange: (query: string) => void;
}
/**
 * Componente CoinSearch (barra de búsqueda para filtrar monedas)
 * @param {CoinSearchProps} props - Props que contienen la función onChange
 * @returns {JSX.Element} El componente renderizado
 */
const CoinSearch: React.FC<CoinSearchProps> = memo(({ onChange }) => {
  const [query, setQuery] = useState<string>('');

  /**
  Cambia el estado de query y devuelve su valor a onChange
  @param {string} queryInput
  */
  const handleText = useCallback((queryInput: string) => {
    setQuery(queryInput);
    if (onChange) {
      onChange(queryInput);
    }
  }, [onChange]);

  return (
    <View testID="coinSearch">
      <TextInput
        testID="input"
        style={[
          styles.textInput,
          Platform.OS == 'ios' ? styles.textInputIos : styles.textInputAndroid,
        ]}
        onChangeText={handleText}
        value={query}
        placeholder="Search coin"
        placeholderTextColor="white"
        accessible={true}
        accessibilityLabel="Search coin"
      />
    </View>
  );
});

const styles = StyleSheet.create({
  textInput: {
    height: 46,
    backgroundColor: colors.blackPearl,
    paddingLeft: 16,
    color: colors.white,
  },
  textInputAndroid: {
    borderWidth: 2,
    borderBottomColor: colors.zircon,
  },
  textInputIos: {
    margin: 8,
    borderRadius: 8,
  },
});

export default CoinSearch;
