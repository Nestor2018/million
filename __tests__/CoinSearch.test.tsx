import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import CoinSearch from '../src/components/coinSearch/CoinSearch';

describe('CoinSearch Component', () => {
  it('should call onChange when text is entered', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(<CoinSearch onChange={mockOnChange} />);
    const mockText = 'bitcoin';

    const input = getByTestId('input');
    fireEvent.changeText(input, mockText);

    expect(mockOnChange).toHaveBeenCalledWith(mockText);
  });

  it('should update the input value when text is entered', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(<CoinSearch onChange={mockOnChange} />);
    const mockText = 'bitcoin';
    const input = getByTestId('input');

    fireEvent.changeText(input, mockText);
    expect(input.props.value).toBe(mockText); // Verifica que el valor del campo sea correcto
  });

  it('should display the correct placeholder', () => {
    const mockOnChange = jest.fn();
    const { getByPlaceholderText } = render(<CoinSearch onChange={mockOnChange} />);

    const placeholder = getByPlaceholderText('Search coin');
    expect(placeholder).toBeTruthy(); // Verifica que el placeholder esté presente
  });

  it('should handle empty input correctly', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(<CoinSearch onChange={mockOnChange} />);
    const input = getByTestId('input');

    fireEvent.changeText(input, ''); // Simula un campo vacío
    expect(mockOnChange).toHaveBeenCalledWith(''); // Verifica que onChange reciba un string vacío
    expect(input.props.value).toBe(''); // Verifica que el valor del campo sea vacío
  });

  it('is accessible with proper accessibility labels', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(<CoinSearch onChange={mockOnChange} />);

    const input = getByTestId('input');
    expect(input.props.accessible).toBe(true); // Verifica que el campo sea accesible
    expect(input.props.accessibilityLabel).toBe('Search coin'); // Verifica el label de accesibilidad
  });

  it('does not re-render when props do not change', () => {
    const mockOnChange = jest.fn();
    const { rerender, getByTestId } = render(<CoinSearch onChange={mockOnChange} />);

    const initialRender = getByTestId('input');
    expect(initialRender).toBeTruthy();

    // Re-renderiza el componente con las mismas props
    rerender(<CoinSearch onChange={mockOnChange} />);
    const afterRerender = getByTestId('input');

    // Verifica que el componente no haya cambiado
    expect(afterRerender).toBe(initialRender);
  });

  it('matches snapshot', () => {
    const mockOnChange = jest.fn();
    const tree = render(<CoinSearch onChange={mockOnChange} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles long input values correctly', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(<CoinSearch onChange={mockOnChange} />);
    const input = getByTestId('input');

    const longText = 'a'.repeat(100); // Simula un texto muy largo
    fireEvent.changeText(input, longText);

    expect(mockOnChange).toHaveBeenCalledWith(longText); // Verifica que onChange reciba el texto largo
    expect(input.props.value).toBe(longText); // Verifica que el valor del campo sea correcto
  });
});
