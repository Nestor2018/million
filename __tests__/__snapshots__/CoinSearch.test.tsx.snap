// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinSearch Component matches snapshot 1`] = `
<View
  testID="coinSearch"
>
  <TextInput
    accessibilityLabel="Search coin"
    accessible={true}
    onChangeText={[Function]}
    placeholder="Search coin"
    placeholderTextColor="white"
    style={
      [
        undefined,
        undefined,
      ]
    }
    testID="input"
    value=""
  />
</View>
`;
